### PR TITLE
add volume api support for convoy

### PR DESCRIPTION
**Pull request by wlan0**
@ibuildthecloud This requires python-agent changes to be implemeneted.

### CODE REVIEW

File: tests/integration/cattletest/core/test_volume.py [hunk 1]
```diff
@@ -11,3 +11,73 @@ def test_volume_delete_active(client, context):
 
     volume = client.wait_success(client.delete(volume))
     assert volume.state == 'removed'
+
+
+def test_volume_create_state(client, context):
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Please test create without volumeDriver param.  It should fail.

Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Test create with options

File: resources/content/schema/base/volume.json [hunk 2]
```diff
@@ -1,8 +1,19 @@
 {
     "resourceFields": {
+        "name": {
+            "unique": true
+        },
         "isHostPath": {
             "type": "boolean",
             "default": false
+        },
+        "volumeDriver": {
```
Comment (ibuildthecloud) [inline resources/content/schema/base/volume.json]: Call this `driver`

File: resources/content/schema/base/volume.json [hunk 3]
```diff
@@ -1,8 +1,19 @@
...
+            "type": "string",
+            "nullable": true
+        },
+        "volumeDriverOpts": {
```
Comment (ibuildthecloud) [inline resources/content/schema/base/volume.json]: `driverOpts`

File: resources/content/schema/base/volume.json [hunk 4]
```diff
@@ -1,8 +1,19 @@
 {
     "resourceFields": {
+        "name": {
```
Comment (ibuildthecloud) [inline resources/content/schema/base/volume.json]: should be required

File: resources/content/schema/base/volume.json (hunk 2 above)
Comment (ibuildthecloud) [inline resources/content/schema/base/volume.json]: should be required

File: tests/integration/cattletest/core/test_volume.py [hunk 5]
```diff
@@ -11,3 +11,73 @@ def test_volume_delete_active(client, context):
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Need tests in test_authorization.py

File: tests/integration/cattletest/core/test_volume.py (hunk 1 above)
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Test volume delete

Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Test container deployed to host where volume is on different host

File: tests/integration/cattletest/core/test_volume.py [hunk 6]
```diff
@@ -11,3 +11,73 @@ def test_volume_delete_active(client, context):
...
+    vol1Found = False
+    vol2Found = False
+
+    for vol in c.volumes():
+        if vol.id == v1.id:
+            vol1Found = True
+        if vol.id == v2.id:
+            vol2Found = True
+
+    assert not vol1Found and not vol2Found
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Need to assert that the volume got allocated to a storagePool.  Look for the volumeStoragePoolMap type using the super_client.

File: resources/content/schema/base/container.json [hunk 7]
```diff
@@ -218,10 +218,14 @@
         "readOnly": {
             "type": "boolean",
             "default": false
-	},
+	    },
```
Comment (ibuildthecloud) [inline resources/content/schema/base/container.json]: remove whitespace change

File: resources/content/cattle-global.properties [hunk 8]
```diff
@@ -8,7 +8,7 @@ api.ui.css.url=https://releases.rancher.com/api-ui/1.0.4/ui.min.css
 agent.instance.image=rancher/agent-instance:v0.4.1
 cluster.image.name=${agent.instance.image}
 bootstrap.required.image=rancher/agent:v0.8.2
-agent.package.python.agent.url=https://github.com/rancher/python-agent/releases/download/v0.53.0/python-agent.tar.gz
+agent.package.python.agent.url=https://s3-us-west-1.amazonaws.com/wlan0/python-agent.tar.gz
```
Comment (ibuildthecloud) [inline resources/content/cattle-global.properties]: That won't work :)

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java [hunk 9]
```diff
@@ -4,12 +4,26 @@
...
 
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang.StringUtils;
+
 public class VolumeManager extends AbstractJooqResourceManager {
 
+    private static final String VOLUME_DRIVER = "volumeDriver";
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java]: These should go in VolumeConstants as FIELD_VOLUME_DRIVER, etc.

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java [hunk 10]
```diff
@@ -20,6 +34,21 @@
...
 
+    @SuppressWarnings("unchecked")
+    @Override
+    protected <T> T createAndScheduleObject(Class<T> clz, Map<String, Object> properties) {
+        String driver = (String) properties.get(VOLUME_DRIVER);
+        String name = (String) properties.get(NAME);
+        if (StringUtils.isNotEmpty(driver) && StringUtils.isNotEmpty(name)) {
+            String uri = StringUtils.join(new String[] { driver, ":///", name });
+            properties.put(URI, uri);
+            properties.put(DEVICE_NUM, 0);
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java]: Why device num 0?  What device number is set for docker volumes today?

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java [hunk 11]
```diff
@@ -20,6 +34,21 @@
         return new Class<?>[] { Volume.class };
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    protected <T> T createAndScheduleObject(Class<T> clz, Map<String, Object> properties) {
+        String driver = (String) properties.get(VOLUME_DRIVER);
+        String name = (String) properties.get(NAME);
+        if (StringUtils.isNotEmpty(driver) && StringUtils.isNotEmpty(name)) {
+            String uri = StringUtils.join(new String[] { driver, ":///", name });
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java]: `String.format("%s:///%s", driver, name)` is a bit more straight forward

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java [hunk 12]
```diff
@@ -4,12 +4,26 @@
...
+import org.apache.commons.lang.StringUtils;
+
 public class VolumeManager extends AbstractJooqResourceManager {
 
+    private static final String VOLUME_DRIVER = "volumeDriver";
+    private static final String NAME = "name";
+    private static final String URI = "uri";
+    private static final String DEVICE_NUM = "deviceNumber"; 
+
+    ObjectManager objectManager;
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java]: This object should already be available to you.  Maybe `getObjectManager()`?

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java [hunk 13]
```diff
@@ -37,7 +37,8 @@
 @Named
 public class InstanceAllocate extends EventBasedProcessHandler {
     private static final Logger log = LoggerFactory.getLogger(InstanceAllocate.class);
-
+    private static final String DATA_VOLUME_MOUNTS = "dataVolumeMounts";
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java]: This should be in VolumeConstants as `FIELD_DATA_VOLUME_MOUNTS`

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java [hunk 14]
```diff
@@ -96,6 +97,7 @@ private void handleCluster(Long requestedHostId, Instance instance, Host cluster
         }
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java]: Needed?

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java [hunk 15]
```diff
@@ -109,12 +111,30 @@ protected HandlerResult postEvent(ProcessState state, ProcessInstance process, M
         }
 
         List<Volume> volumes = getObjectManager().children(instance, Volume.class);
+        Map<String, Object> dataVolumeMounts = DataAccessor.field(instance, DATA_VOLUME_MOUNTS, Map.class);
+
+        if (dataVolumeMounts != null) {
+            for (Map.Entry<String, Object> entry : dataVolumeMounts.entrySet()) {
+                Volume v = getObjectManager().loadResource(Volume.class, ((Integer) entry.getValue()).longValue());
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java]: cast to `Number` or you might get class cast exceptions here.

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java [hunk 16]
```diff
@@ -109,12 +111,30 @@ protected HandlerResult postEvent(ProcessState state, ProcessInstance process, M
...
+                }
+            }
+        }
 
         for (Volume v : volumes) {
             allocate(v, state.getData());
         }
 
         volumes = getObjectManager().children(instance, Volume.class);
+        if (dataVolumeMounts != null) {
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java]: This feels like deja vu, what is the difference from this block and the one above?

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 17]
```diff
@@ -56,6 +57,9 @@
     @Inject
     LabelsDao labelsDao;
 
+    private static String DATA_VOLUME_MOUNTS = "dataVolumeMounts";
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: Move to VolumeConstants

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 18]
```diff
@@ -87,6 +92,30 @@ protected boolean shouldStart(Instance instance) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private List<Volume> processManagedVolumes(ProcessState state, Instance instance) {
+        List<Volume> volumes = new ArrayList<Volume>();
+        Map<String, Object> dataVolumeMounts = DataAccessor.field(instance, DATA_VOLUME_MOUNTS, Map.class);
+        if (dataVolumeMounts != null) {
+            List<String> dataVolumes = (List<String>) DataAccessor.field(instance, DATA_VOLUMES, Map.class);
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: You're casting a Map to a List

File: tests/integration/cattletest/core/test_volume.py (hunk 1 above)
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Test where you pass in dataVolumes and dataVolumesMounts

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 19]
```diff
@@ -87,6 +92,30 @@ protected boolean shouldStart(Instance instance) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private List<Volume> processManagedVolumes(ProcessState state, Instance instance) {
+        List<Volume> volumes = new ArrayList<Volume>();
+        Map<String, Object> dataVolumeMounts = DataAccessor.field(instance, DATA_VOLUME_MOUNTS, Map.class);
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: Use the `DataAccessor.fieldMap()` and `fieldString` and `fieldStringList`, etc.  Less casting madness

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 20]
```diff
@@ -87,6 +92,30 @@ protected boolean shouldStart(Instance instance) {
...
+        if (dataVolumeMounts != null) {
+            List<String> dataVolumes = (List<String>) DataAccessor.field(instance, DATA_VOLUMES, Map.class);
+            if (dataVolumes == null) {
+                dataVolumes = new ArrayList<String>();
+            }
+            for (Map.Entry<String, Object> mountedVols : dataVolumeMounts.entrySet()) {
+                Long volId = ((Integer) mountedVols.getValue()).longValue();
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: If needed always cast to a Number

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceStart.java [hunk 21]
```diff
@@ -136,7 +137,7 @@ protected int getMaxComputeTries(Instance instance) {
     }
 
     protected HandlerResult stopOrRemove(ProcessState state, Instance instance, ExecutionException e) {
-        if (InstanceCreate.isCreateStart(state) && !ContainerEventCreate.isNativeDockerStart(state) ) {
+        if (InstanceCreate.isCreateStart(state) && !ContainerEventCreate.isNativeDockerStart(state)) {
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceStart.java]: white space change, remove

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceStart.java [hunk 22]
```diff
@@ -166,9 +167,21 @@ protected void allocate(Instance instance) {
         execute("instance.allocate", instance, null);
     }
 
+    @SuppressWarnings("unchecked")
     protected void storage(Instance instance, ProcessState state) {
         List<Volume> volumes = getObjectManager().children(instance, Volume.class);
 
+        Map<String, Object> dataVolumeMounts = DataAccessor.field(instance, DATA_VOLUME_MOUNTS, Map.class);
+
+        if (dataVolumeMounts != null) {
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceStart.java]: Why don't you create a util method for this,  I think this is the 3rd time I'm seen this code.

File: code/iaas/logic/src/main/java/io/cattle/platform/process/volume/VolumeActivate.java [hunk 23]
```diff
@@ -30,10 +30,9 @@
 
     @Override
     public HandlerResult handle(ProcessState state, ProcessInstance process) {
-        Volume volume = (Volume) state.getResource();
-
-        Set<Long> pools = new HashSet<Long>();
+        Volume volume = (Volume) state.getResource();        Set<Long> pools = new HashSet<Long>();
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/volume/VolumeActivate.java]: Did something actually change here?

File: code/implementation/docker/common/src/main/java/io/cattle/platform/docker/transform/DockerInspectTransformVolume.java [hunk 24]
```diff
@@ -6,13 +6,15 @@
     String hostPath;
     String accessMode;
     boolean bindMount;
-    
-    public DockerInspectTransformVolume(String cp, String hp, String am, boolean bm) {
+    String driver;
+
+    public DockerInspectTransformVolume(String cp, String hp, String am, boolean bm, String dr) {
```
Comment (ibuildthecloud) [inline code/implementation/docker/common/src/main/java/io/cattle/platform/docker/transform/DockerInspectTransformVolume.java]: You need to add a python test for the transformer in `test_tranform_inspect`

File: code/implementation/docker/compute/src/main/java/io/cattle/platform/docker/process/instancehostmap/DockerPostInstanceHostMapActivate.java [hunk 25]
```diff
@@ -167,7 +169,13 @@ protected void processVolumes(Instance instance, Host host, ProcessState state)
         }
 
         for (DockerInspectTransformVolume dVol : dockerVolumes) {
-            String volumeUri = String.format(VolumeConstants.URI_FORMAT, dVol.getHostPath());
+            String volumeUri = null;
+            String driver = dVol.getDriver();
+            if (StringUtils.isEmpty(driver) || StringUtils.equals(driver, "local")) {
+                volumeUri = String.format(VolumeConstants.URI_FORMAT, dVol.getHostPath());
+            } else {
+                volumeUri = dVol.getHostPath();
```
Comment (ibuildthecloud) [inline code/implementation/docker/compute/src/main/java/io/cattle/platform/docker/process/instancehostmap/DockerPostInstanceHostMapActivate.java]: What is this value?  Is this `driver:///name`?

File: tests/integration/cattletest/core/test_volume.py (hunk 1 above)
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Need to add tests in `test_docker.py` that the back populating from the docker inspect works.

**Comment by ibuildthecloud** [comment]
Can you rebase?

File: code/implementation/docker/compute/src/main/java/io/cattle/platform/docker/process/instancehostmap/DockerPostInstanceHostMapActivate.java (hunk 25 above)
Comment (wlan0) [inline code/implementation/docker/compute/src/main/java/io/cattle/platform/docker/process/instancehostmap/DockerPostInstanceHostMapActivate.java]: yes

File: code/implementation/docker/common/src/main/java/io/cattle/platform/docker/transform/DockerInspectTransformVolume.java (hunk 24 above)
Comment (wlan0) [inline code/implementation/docker/common/src/main/java/io/cattle/platform/docker/transform/DockerInspectTransformVolume.java]: After disucussing offline with @ibuildthecloud, No tests are needed for this.

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java [hunk 26]
```diff
@@ -8,8 +8,16 @@
...
 
+import java.util.Map;
+
+import org.apache.commons.lang.StringUtils;
+
 public class VolumeManager extends AbstractJooqResourceManager {
 
+    private static final String NAME = "name";
+    private static final String URI = "uri";
+    private static final String DEVICE_NUM = "deviceNumber";
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java]: Use io.cattle.platform.object.meta.ObjectMetaDataManager.NAME_FIELD

Move to VolumeConstants.FIELD_URI, etc...

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java [hunk 27]
```diff
@@ -20,6 +28,21 @@
         return new Class<?>[] { Volume.class };
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    protected <T> T createAndScheduleObject(Class<T> clz, Map<String, Object> properties) {
+        String driver = (String) properties.get(VolumeConstants.FIELD_VOLUME_DRIVER);
+        String name = (String) properties.get(NAME);
+        if (StringUtils.isNotEmpty(driver) && StringUtils.isNotEmpty(name)) {
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java]: If driver set and name not return send validation error

File: code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java [hunk 28]
```diff
@@ -20,6 +28,21 @@
...
+            String uri = String.format("%s:///%s", driver, name);
+            properties.put(URI, uri);
+            properties.put(DEVICE_NUM, -1);
+        }
+        Object result = getObjectManager().create(clz, properties);
```
Comment (ibuildthecloud) [inline code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/manager/VolumeManager.java]: just return don't use temp variable

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java [hunk 29]
```diff
@@ -53,9 +54,11 @@ public InstanceAllocate() {
 
     @Override
     public HandlerResult handle(ProcessState state, ProcessInstance process) {
-        // Check resourceType for instance and whether the host is a cluster or not.
-        // If resourceType is a cluster (host), do not bother with normal allocation process.
-        final Instance instance = (Instance)state.getResource();
+        // Check resourceType for instance and whether the host is a cluster or
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java]: Don't change

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java [hunk 30]
```diff
@@ -77,9 +80,7 @@ private void handleCluster(Long requestedHostId, Instance instance, Host cluster
 
         // create instanceHostMap binding with cluster.
         if (mapDao.findNonRemoved(InstanceHostMap.class, Host.class, requestedHostId, Instance.class, instance.getId()) == null) {
-            objectManager.create(InstanceHostMap.class,
-                    INSTANCE_HOST_MAP.HOST_ID, requestedHostId,
-                    INSTANCE_HOST_MAP.INSTANCE_ID, instance.getId());
+            objectManager.create(InstanceHostMap.class, INSTANCE_HOST_MAP.HOST_ID, requestedHostId, INSTANCE_HOST_MAP.INSTANCE_ID, instance.getId());
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java]: Don't change formatting

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 31]
```diff
@@ -1,10 +1,9 @@
 package io.cattle.platform.process.instance;
 
-import static io.cattle.platform.core.model.tables.CredentialInstanceMapTable.*;
-import static io.cattle.platform.core.model.tables.NetworkTable.*;
-import static io.cattle.platform.core.model.tables.NicTable.*;
-import static io.cattle.platform.core.model.tables.VolumeTable.*;
-
+import static io.cattle.platform.core.model.tables.CredentialInstanceMapTable.CREDENTIAL_INSTANCE_MAP;
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: In eclipse Java -> Code Style -> Organize Imports, set .* to 1

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 32]
```diff
@@ -88,6 +89,34 @@ protected boolean shouldStart(Instance instance) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private List<Volume> processManagedVolumes(ProcessState state, Instance instance) {
+        List<Volume> volumes = new ArrayList<Volume>();
+        Map<String, Object> dataVolumeMounts = DataAccessor.field(instance, InstanceConstants.FIELD_DATA_VOLUME_MOUNTS, Map.class);
+        if (dataVolumeMounts != null) {
+            List<String> dataVolumes = DataAccessor.field(instance, InstanceConstants.FIELD_DATA_VOLUMES, List.class);
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: Can you early exist, so just return volumes and not have the whole block indented.

Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: Use fieldString method

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 33]
```diff
@@ -88,6 +89,34 @@ protected boolean shouldStart(Instance instance) {
...
+            if (dataVolumes == null) {
+                dataVolumes = new ArrayList<String>();
+            }
+            for (Map.Entry<String, Object> mountedVols : dataVolumeMounts.entrySet()) {
+                Long volId = ((Number) mountedVols.getValue()).longValue();
+
+                Volume vol = objectManager.loadResource(Volume.class, volId);
+                if (vol != null) {
+                    volumes.add(vol);
+                    String volDescriptor = vol.getName() + ":" + mountedVols.getKey();
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: Strip key()

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 34]
```diff
@@ -88,6 +89,34 @@ protected boolean shouldStart(Instance instance) {
...
+                    if (!dataVolumes.contains(volDescriptor)) {
+                        dataVolumes.add(volDescriptor);
+                    }
+                }
+            }
+            Map<String, Object> data = new HashMap<String, Object>();
+            data.put(InstanceConstants.FIELD_DATA_VOLUMES, dataVolumes);
+            state.applyData(data);
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: Return this from handle(), don't set in the middle of the handler.

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java [hunk 35]
```diff
@@ -154,13 +183,13 @@ protected Volume getRoot(Instance instance, List<Volume> volumes) {
...
         for (Volume volume : volumes) {
-            if (volume.getDeviceNumber().intValue() == 0) {
+            if (volume.getDeviceNumber() != null && volume.getDeviceNumber().intValue() == 0) {
                 return volume;
             }
         }
 
-        return objectManager.create(Volume.class, VOLUME.ACCOUNT_ID, instance.getAccountId(), VOLUME.INSTANCE_ID, instance.getId(), VOLUME.IMAGE_ID, instance
-                .getImageId(), VOLUME.DEVICE_NUMBER, 0, VOLUME.ATTACHED_STATE, CommonStatesConstants.ACTIVE);
+        return objectManager.create(Volume.class, VOLUME.ACCOUNT_ID, instance.getAccountId(), VOLUME.INSTANCE_ID, instance.getId(), VOLUME.IMAGE_ID,
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceCreate.java]: whitespace change?

File: code/iaas/logic/src/main/java/io/cattle/platform/process/util/InstanceHelpers.java [hunk 36]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class InstanceHelpers {
+
+    public static List<Volume> extractVolumesFromMounts(Instance instance, ObjectManager objectManager) {
+        List<Volume> volumes = new ArrayList<Volume>();
+        @SuppressWarnings("unchecked")
+        Map<String, Object> dataVolumeMounts = DataAccessor.field(instance, InstanceConstants.FIELD_DATA_VOLUME_MOUNTS, Map.class);
```
Comment (ibuildthecloud) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/util/InstanceHelpers.java]: use fieldMap()

File: resources/content/schema/user/user-auth.json [hunk 37]
```diff
@@ -295,11 +296,14 @@
         "userPreference" : "crud",
         "userPreference.value" : "crud",
 
-        "volume" : "r",
+        "volume" : "crd",
```
Comment (ibuildthecloud) [inline resources/content/schema/user/user-auth.json]: should be "r"

File: tests/integration/cattletest/core/test_container.py [hunk 38]
```diff
@@ -75,6 +75,8 @@ def test_container_create_only(super_client, client, context):
     })
 
     volumes = container.volumes()
+    print container
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_container.py]: remove print

File: tests/integration/cattletest/core/test_docker.py [hunk 39]
```diff
@@ -450,6 +450,105 @@ def test_no_port_override(docker_client, super_client):
 
 
 @if_docker
+def test_docker_managed_volumes(docker_client, super_client):
+
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_docker.py]: delete line

File: tests/integration/cattletest/core/test_docker.py [hunk 40]
```diff
@@ -450,6 +450,105 @@ def test_no_port_override(docker_client, super_client):
...
+    baz_host_path = '/tmp/baz%s' % bind_mount_uuid
+    baz_bind_mount = '%s:/baz:ro' % baz_host_path
+
+    volName = 'test-%s' % random_str()
+    mv1 = docker_client.create_volume(name=volName,
+                                      driver='local')
+    mv1 = docker_client.wait_success(mv1)
+
+    managed_mnt_pt = '/tmp/managedbad%s' % bind_mount_uuid
+    dvms = {managed_mnt_pt: mv1.id}
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_docker.py]: Add a second managed volume

File: tests/integration/cattletest/core/test_volume.py [hunk 41]
```diff
@@ -3,7 +3,7 @@
 
 def test_volume_delete_active(client, context):
     c = client.create_container(imageUuid=context.image_uuid)
-    c = client.wait_success(c)
+    c = client.wait_success(c, timeout=240)
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Why timeout?

File: tests/integration/cattletest/core/test_volume.py [hunk 42]
```diff
@@ -11,3 +11,104 @@ def test_volume_delete_active(client, context):
...
     assert volume.state == 'removed'
+
+
+def test_volume_create_state(client, context):
+    name = random_str()
+    c = client.create_volume(name=name, driver='local')
+    c = client.wait_success(c)
+    assert c.state == 'requested'
+
+    assert c.uri == 'local:///%s' % name
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: assert driverOpts and driver values

File: tests/integration/cattletest/core/test_volume.py [hunk 43]
```diff
@@ -11,3 +11,104 @@ def test_volume_delete_active(client, context):
...
+
+def test_volume_create_with_opts(client, context):
+    name = random_str()
+    c = client.create_volume(name=name,
+                             driver='local',
+                             driverOpts={'size': '1G'})
+    c = client.wait_success(c)
+    assert c.state == 'requested'
+
+    assert c.uri == 'local:///%s' % name
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: assert driverOpts and driver values

File: tests/integration/cattletest/core/test_volume.py [hunk 44]
```diff
@@ -11,3 +11,104 @@ def test_volume_delete_active(client, context):
...
+    v1 = client.wait_success(v1)
+    v2 = client.wait_success(v2)
+
+    assert v1.state == 'active'
+    assert v2.state == 'active'
+
+    vol1Found = False
+    vol2Found = False
+
+    for vol in c.volumes():
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: more readable if you do

```python
assert len(c.volumes()) == 1
assert c.volumes()[0].id not in [v1.id, v2.id]
```

File: tests/integration/cattletest/core/test_volume.py [hunk 45]
```diff
@@ -11,3 +11,104 @@ def test_volume_delete_active(client, context):
...
+        if vol.id == v2.id:
+            vol2Found = True
+
+    assert not vol1Found and not vol2Found
+
+    vsp1 = super_client.list_volumeStoragePoolMap(volumeId=v1.id)
+    vsp2 = super_client.list_volumeStoragePoolMap(volumeId=v2.id)
+
+    assert vsp1 is not None
+    assert vsp2 is not None
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Assert vsp1, vsp1 are mapped to the correct host

Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: assert len(vsp1) == 1

File: tests/integration/cattletest/core/test_volume.py [hunk 46]
```diff
@@ -11,3 +11,104 @@ def test_volume_delete_active(client, context):
...
+    c = client.wait_success(c)
+    assert c.state == 'requested'
+
+    assert c.uri == 'local:///%s' % name
+
+    volume = client.wait_success(client.delete(c))
+    assert volume.state == 'removed'
+
+
+def test_create_container_with_volume(client, context, super_client):
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: pass in `new_context, super_client`
set `client = new_context.client`

File: tests/integration/cattletest/core/test_volume.py [hunk 47]
```diff
@@ -11,3 +11,104 @@ def test_volume_delete_active(client, context):
...
+
+    assert not vol1Found and not vol2Found
+
+    vsp1 = super_client.list_volumeStoragePoolMap(volumeId=v1.id)
+    vsp2 = super_client.list_volumeStoragePoolMap(volumeId=v2.id)
+
+    assert vsp1 is not None
+    assert vsp2 is not None
+
+    new_host = super_client.create_host()
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: `new_host = register_simulated_host(new_context.client)`

File: tests/integration/cattletest/core/test_volume.py [hunk 48]
```diff
@@ -11,3 +11,104 @@ def test_volume_delete_active(client, context):
...
+    new_host = super_client.wait_success(new_host)
+
+    try:
```
Comment (ibuildthecloud) [inline tests/integration/cattletest/core/test_volume.py]: Use this style

```python
    with pytest.raises(ApiError) as e:
       blah...
    
    assert e.value.error.status == 422
    assert e.value.error.code == 'NotUnique'
    assert e.value.error.fieldName == 'name'
```

**Pull request merged**